### PR TITLE
MOSO: Combining SOAP and Muon 

### PR DESCRIPTION
**Pull request by mkhona-nvidia**
# MOSO

MOSO, short for Momentum One-Sided SOAP, combines Muon-style momentum with SOAP's eigenbasis Adam update, but keeps the preconditioner one-sided on the smaller matrix dimension. For a momentum matrix \(M_t\), MOSO accumulates a SOAP-style covariance over momentum instead of raw gradients, using $(C_t = \beta_s C_{t-1} + (1 - \beta_s) M_t M_t^T\)$ for the left-preconditioned case, or $(C_t = \beta_s C_{t-1} + (1 - \beta_s) M_t^T M_t\)$ for the right-preconditioned case. With $(C_t = Q_M \Lambda_M Q_M^T\)$, the left-side update is

```math
U_t = Q_M \text{Adam}(Q_M^T M_t),
```

with the analogous right-side projection $(U_t = \text{Adam}(M_t Q_M) Q_M^T\)$ when the column dimension is smaller. This can be read as one-sided SOAP on Muon momentum: rotate $(M_t\)$ into the momentum-covariance eigenbasis, run the inner Adam update there, and rotate back.

### CODE REVIEW

File: emerging_optimizers/soap/moso.py [hunk 1]
```diff
@@ -0,0 +1,337 @@
...
+        qr_fp32_matmul_prec: Precision of the matmul operations in QR decomposition.
+        power_iter_steps: Number of power iteration steps to perform before QR decomposition.
+        scale_mode: Muon update scale mode.
+        extra_scale_factor: Additional update scale factor.
+        max_update_rms: Clip the update RMS to this value (0 means no clipping).
+    """
+
+    def __init__(
+        self,
+        params: ParamsT,
```
Comment (skyw) [inline emerging_optimizers/soap/moso.py:67]: I think the argument list is too long. No need to have all of them for experimental purpose. Can add them when we actually want to shmoo them. 

I think we can at least remove
- weight_decay_method
- nesterov
- correct_bias
- correct_shampoo_beta_bias
- fp32_matmul_prec
- use_eigh: bool = False,
- qr_fp32_matmul_prec
- power_iter_steps

File: emerging_optimizers/soap/moso.py [hunk 2]
```diff
@@ -0,0 +1,337 @@
...
+def _update_one_sided_momentum_factor(
+    momentum_factor: torch.Tensor,
+    momentum: torch.Tensor,
+    shampoo_beta: float,
+) -> None:
+    """Update the smaller-side covariance of the Muon momentum."""
+    if momentum.shape[0] <= momentum.shape[1]:
+        momentum_factor.lerp_(momentum @ momentum.T, 1 - shampoo_beta)
+    else:
+        momentum_factor.lerp_(momentum.T @ momentum, 1 - shampoo_beta)
```
Comment (skyw) [inline emerging_optimizers/soap/moso.py]: DRY.
```python
    maybe_transposed_m = momentum.mT if momentum.shape[0] > momentum.shape[1] else momentum
    momentum_factor.lerp_(maybe_transposed_m @ maybe_transposed_m .T, 1 - shampoo_beta)
```
or something similar is easy to read. maybe also name the condition `left_preconditioned` to be consistent.

File: emerging_optimizers/soap/moso.py [hunk 3]
```diff
@@ -0,0 +1,337 @@
...
+    )
+
+    if use_eigh:
+        updated_eigenbasis = soap_utils.get_eigenbasis_eigh([momentum_factor])[0]
+    else:
+        updated_eigenbasis = soap_utils.get_eigenbasis_qr(
+            [momentum_factor],
+            [eigenbasis],
+            power_iter_steps=power_iter_steps,
+        )[0]
```
Comment (skyw) [inline emerging_optimizers/soap/moso.py]: unpact instead of indexing. i.e. "updated_eigenbasis, _ = ".

File: emerging_optimizers/soap/sh_muon.py [hunk 4]
```diff
@@ -0,0 +1,337 @@
...
+@torch.no_grad()  # type: ignore[misc]
+def _sort_one_sided_eigenbasis_and_exp_avg_sq(
+    momentum_factor: torch.Tensor,
+    eigenbasis: torch.Tensor,
+    exp_avg_sq: torch.Tensor,
+    left_preconditioned: bool,
+) -> tuple[torch.Tensor, torch.Tensor]:
+    """Sort eigenbasis slots by approximate eigenvalue and permute Adam second moments."""
+    approx_eigvals = utils.eig.conjugate(momentum_factor, eigenbasis, diag=True)
+    sort_idx = torch.argsort(approx_eigvals, descending=True)
```
Comment (skyw) [inline emerging_optimizers/soap/sh_muon.py]: I think we should add stable=True here to be extra cautious.

File: tests/test_sh_muon.py [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+            projected = grad @ eigenbasis
+            adam_projected = projected / (projected.abs() + optimizer.param_groups[0]["eps"])
+            expected_update = adam_projected @ eigenbasis.T
+
+        expected_update = expected_update * get_muon_scale_factor(*shape, mode="spectral")
+        applied_update = -param.detach() / lr
+        torch.testing.assert_close(
+            applied_update,
+            expected_update,
+            atol=1e-4,
```
Comment (skyw) [inline tests/test_sh_muon.py]: would the update be close to 0? if not, set atol to 0.
Also "matches" in test name implies atol=rtol=0, otherwise use closes to

**Review by skyw — CHANGES_REQUESTED** [review]
Other than bit DRY, which is not this PR's problem, mostly ok. will approve after.

**Review by skyw — APPROVED** [review]
